rfcs/proposed/numa_support/create-numa-arenas.md

```diff
@@ -0,0 +1,110 @@
+# API to simplify creation of task arenas constrained to NUMA nodes
+
+This sub-RFC proposes an API to ease creation of a one-per-NUMA-node set of task arenas.
+
+## Introduction
+
+The code example in the [overarching RFC for NUMA support](README.md) shows the likely
+pattern of using task arenas to distribute computation across all NUMA domains on a system.
+Let's take a closer look at the part where arenas are created and initialized.
+
+```c++
+  std::vector<tbb::numa_node_id> numa_nodes = tbb::info::numa_nodes();
+  std::vector<tbb::task_arena> arenas(numa_nodes.size());
+  std::vector<tbb::task_group> task_groups(numa_nodes.size());
+
+  // initialize each arena, each constrained to a different NUMA node
+  for (int i = 0; i < numa_nodes.size(); i++)
+    arenas[i].initialize(tbb::task_arena::constraints(numa_nodes[i]), 0);
+```
+
+The first line obtains a vector of NUMA node IDs for the system. Then, a vector of the same size
+is created to store `tbb::task_arena` objects, each constrained to one of the NUMA nodes.
+Another vector holds `task_group` instances used later to submit and wait for completion
+of the work in each of the arenas - it is necessary because `task_arena` does not provide
+any work synchronization API. Finally, the loop over all NUMA nodes initializes associated
+task arenas with proper constraints.
+
+While not incomprehensible, the code is quite verbose and arguably too explicit for the typical scenario
+of creating a set of arenas across all available NUMA domains. There is also risk of subtle issues.
+The default constructor of `task_arena` reserves a slot for an application thread. The arena initialization
+at the last line explicitly overwrites it to 0 to allow TBB worker threads taking all the slots, however
+this nuance might be unknown and easy to miss, potentially resulting in underutilization of CPU resources.
+
+## Proposal
+
+We propose to introduce a special function to create the set of task arenas, one per NUMA node on the system.
+The initialization code equivalent to the example above would be:
+
+```c++
+  std::vector<tbb::task_arena> arenas = tbb::create_numa_task_arenas();
+  std::vector<tbb::task_group> task_groups(arenas.size());
+```
+
+### Public API
+
+The function has the following signature:
+
+```c++
+// Defined in tbb/task_arena.h
+
+namespace tbb {
+    std::vector<tbb::task_arena> create_numa_task_arenas(
+        task_arena::constraints other_constraints = {},
+        unsigned reserved_slots = 0
```

Comment:
You mention priorities later, but why _not_ have a default normal priority as an argument? Is there a reason to exclude it? It seems like it would naturally fall after reserved_slots as is usual for task_arena constructors?

Reply:
At this point, I do not want to complicate the API with more "syntax sugar" than practically necessary. I can see how setting other constraints (max concurrency and threads-per-core) can be useful, and I can see a scenario where NUMA arenas are served by different application threads to provide an option for reservation. However I do not easily see the case for changing the priority of all NUMA arenas; so why providing an option for that?

There is also the experimental API that adds thread leave policy as another arena construction option; should it also be a parameter to the arena set factory? And maybe we later decide to combine these options into a structure similar to arena constraints. So I would better wait now and add more overloads later, if there is demand.